src/inmanta/agent/agent.py

```diff
@@ -1332,52 +1490,81 @@ async def on_disconnect(self) -> None:
         for agent_instance in self._instances.values():
             agent_instance.pause("Connection to server lost")
 
-    async def ensure_code(self, environment: uuid.UUID, version: int, resource_types: Sequence[str]) -> set[str]:
-        """Ensure that the code for the given environment and version is loaded"""
+    async def get_code(self, environment: uuid.UUID, version: int, resource_types: Sequence[str]) -> list[ResourceInstallSpec]:
         failed_to_load: set[str] = set()
         if self._loader is None:
             return failed_to_load
 
         # store it outside the loop, but only load when required
         pip_config: Optional[PipConfig] = None
 
-        for rt in set(resource_types):
-            # only one logical thread can load a particular resource type at any time
-            async with self._resource_loader_lock.get(rt):
-                # stop if the last successful load was this one
-                # The combination of the lock and this check causes the reloads to naturally 'batch up'
-                if self._last_loaded[rt] == version:
-                    LOGGER.debug("Code already present for %s version=%d", rt, version)
-                    continue
-                # clear cache, for retry on failure
-                self._last_loaded[rt] = -1
+        resource_install_specs: list[ResourceInstallSpec] = []
+        for resource_type in set(resource_types):
 
-                result: protocol.Result = await self._client.get_source_code(environment, version, rt)
+            if self._last_loaded[resource_type] == version:
+                LOGGER.debug("Code already present for %s version=%d", resource_type, version)
+                continue
```

Comment:
I'm not entirely certain that this is correct now. We're only fetching the code here. The `ensure_code` method that now populates this cache is in-process while it might be that we're fetching code for an external executor.

Reply:
As it is now, I would simply cache the set of `ResourceInstallSpec`  for a given `(resourcetype, version)`